### ISSUE TITLE
scipy.spatial.transform.Rotation

### ISSUE THREAD

**Issue by cpgoodri** [google/jax#3192]
Is there any plan for implementing scipy.spatial.transform.Rotation (or any other 3d rotation package that includes quaternions, Euler angles, and rotation matrices) in the near future? 

I need something along these lines... I'm considering going through the transformations.py package and updating it to be JAX-compatible (probably with only partial functionality), but obviously I don't want to do this if something is in the works here!

**Comment by jakevdp** [comment]
I don't know of any plans to implement those. Feel free to send some pull requests!

**Comment by shoyer** [comment]
One thing to watch out for is that Rotation is a class based interface. These sometimes be can be a bit tricky to wrap in JAX.

Fortunately the Rotation API isn’t intrinsically stateful. It appears to be a thin wrapper around a 1D array of 4D vectors representing quaternions, stored as a 2D array.

We have roughly two options:

1. It might make sense to define Rotation as a “pytree” so JAX’s transformations are defined with respect to its parameters. But implies some potentially strange behavior:
    - From `grad` you could end up with a “Rotation” of gradients, on which Rotation methods may not be well defined.
    - From `vmap’ you could end up with higher dimensional arrays of quaternions, on which methods also might not work properly.
2. The alternative is to not worry about any of this, and require users to pass around arrays, and only use Rotation *inside* JAX transformed functions. This might be slightly limiting, but in this case all transformations are perfectly well defined because they only applied to pure functions of arrays.

Option (2) seems like it’s probably the safer choice. It would also be interesting to know if this matches what’s done for Distribution objects from TensorFlow Probability when used with JAX.

**Comment by cpgoodri** [comment]
Thanks for your replies! I'll admit I don't understand the internals of JAX well enough to be able to comment on the two options. In terms of an api, there is definitely a benefit to being consistent with what scipy already has, but I am personally a fan of the approach that combines all 3d transformations (rotations, translations, shears, reflections, scalings, etc.) in a 4x4 matrix that can be applied to [x,y,z,1] [see here](http://www.it.hiof.no/~borres/j3d/math/threed/p-threed.html). 

I can't commit to anything right now, but I would be interested in being involved at some level on this front.

Anyways, I have a related but different question. Lets say I finish updating the [transformations.py](https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/_gohlketransforms.py) package to be JAX-compatible and test it to the point where I want to make it public. Is there a collection or repository anywhere that lists 3rd party packages that are built to be JAX-compatible? Would something like this be useful as a means of expanding the JAX ecosystem?

**Comment by rdaems** [comment]
If this is useful for anyone, I wrote a `jit`able rotation matrix to quaternion function, based on the function in scipy:

```python
@jit
def matrix_to_quaternion(matrix):
    m = matrix.conj().transpose() # This method assumes row-vector and postmultiplication of that vector

    b0 = jnp.where(m[2, 2] < 0, 1., -1.)
    b1 = jnp.where(b0 * m[0, 0] > m[1, 1], 1., -1.)
    t = 1 + b0 * b1 * m[0, 0] - b1 * m[1, 1] - b0 * m[2, 2]
    
    q = jnp.where(
        b0 > 0,
        jnp.where(
            b1 > 0,
            jnp.array([m[1, 2]-m[2, 1],  t,  m[0, 1]+m[1, 0],  m[2, 0]+m[0, 2]]),
            jnp.array([m[2, 0]-m[0, 2],  m[0, 1]+m[1, 0],  t,  m[1, 2]+m[2, 1]])
        ),
        jnp.where(
            b1 > 0,
            jnp.array([m[0, 1]-m[1, 0],  m[2, 0]+m[0, 2],  m[1, 2]+m[2, 1],  t]),
            jnp.array([t,  m[1, 2]-m[2, 1],  m[2, 0]-m[0, 2],  m[0, 1]-m[1, 0]])
        )
    )
    q *= 0.5 / jnp.sqrt(t)
    return q
```

**Comment by cpgoodri** [comment]
Thanks rdaems.

I forgot to post here that I finished implementing the transformations.py package to be jit-able. See here: https://github.com/cpgoodri/jax_transformations3d

It is actually not 100% implemented, but there is lots of functionality for using quaternious/euler angles/ rotation matrices, etc. Hope someone finds it helpful.

**Comment by harsh306** [comment]
We should also think about  N-dimensional rotations, also, in terms of Euler angles it can be done only in 2d subspace I guess, but maybe someone with geometric algebra background can help? (https://bivector.net/)

**Comment by chrisflesher** [comment]
Here is some code I've been using that implements a `Rotation` class by wrapping the `jax_transformations3d` library. Maybe it would be useful to someone else:

```
import functools
import typing

import jax
import jax.numpy as jnp
import jax_transformations3d


class Rotation(typing.NamedTuple):
    """Rotation in 3 dimensions."""

    quat: jax.Array

    @classmethod
    def from_quat(cls, quat: jax.Array):
        """Initialize from quaternions."""
        return cls(quat)

    @classmethod
    def from_matrix(cls, matrix: jax.Array):
        """Initialize from rotation matrix."""
        return cls(jnp.roll(jax_transformations3d.quaternion_from_matrix(matrix), -1))

    @classmethod
    def from_rotvec(cls, rotvec: jax.Array, degrees: bool = False):
        """Initialize from rotation vectors."""
        norm = jax_transformations3d.vector_norm(rotvec)
        axis = jnp.where(norm > 0, rotvec / norm, jnp.array([1., 0., 0.]))
        angle = jnp.where(degrees, jnp.radians(norm), norm)
        return cls(jnp.roll(jax_transformations3d.quaternion_about_axis(angle, axis), -1))

    @classmethod
    def from_euler(cls, seq: str, angles: jax.Array, degrees: bool = False):
        """Initialize from Euler angles."""
        assert seq == 'xyz'
        assert angles.size == 3
        ai = angles[0]
        aj = angles[1]
        ak = angles[2]
        return cls(jnp.roll(jax_transformations3d.quaternion_from_euler(ai, aj, ak, axes='sxyz'), -1))

    @jax.jit
    def as_quat(self) -> jax.Array:
        """Represent as quaternions."""
        return self.quat

    @jax.jit
    def as_matrix(self) -> jax.Array:
        """Represent as rotation matrix."""
        return jax_transformations3d.quaternion_matrix(jnp.roll(self.quat, 1))[:3, :3]

    @functools.partial(jax.jit, static_argnames=['degrees'])
    def as_rotvec(self, degrees: bool = False) -> jax.Array:
        """Represent as rotation vectors."""
        angle = 2 * jnp.arccos(self.quat[-1])
        wrapped_angle = jnp.where(angle >= jnp.pi, angle - 2*jnp.pi, angle)
        norm = jnp.linalg.norm(self.quat[0:3])
        axis = self.quat[0:3]
        return jnp.where(norm > 0, wrapped_angle * axis / norm, jnp.zeros(3))

    @functools.partial(jax.jit, static_argnames=['seq', 'degrees'])
    def as_euler(self, seq: str, degrees: bool = False):
        """Represent as Euler angles."""
        assert seq == 'xyz'
        ai, aj, ak = jax_transformations3d.euler_from_quaternion(jnp.roll(self.quat, 1), axes='sxyz')
        angles = jnp.array([ai, aj, ak])
        return jnp.where(degrees, jnp.degrees(angles), angles)

    @functools.partial(jax.jit, static_argnames=['inverse'])
    def apply(self, vector: jax.Array, inverse: bool = False) -> jax.Array:
        """Apply this rotation to a set of vectors."""
        matrix = self.as_matrix()
        if inverse:
            result = jnp.einsum('kj,k->j', matrix, vector)
        else:
            result = jnp.einsum('jk,k->j', matrix, vector)
        return result

    @jax.jit
    def __mul__(self, other):
        """Compose this rotation with the other."""
        temp = jax_transformations3d.quaternion_multiply(jnp.roll(self.quat, 1), jnp.roll(other.quat, 1))
        return Rotation.from_quat(jnp.roll(temp, -1))

    @jax.jit
    def inv(self):
        """Invert this rotation."""
        return Rotation.from_quat(jnp.array([self.quat[0], self.quat[1], self.quat[2], -self.quat[3]]))
```

**Comment by chrisflesher** [comment]
Attempted to implement this, see PR #15465.
